spacy/pipeline/edit_tree_lemmatizer.py

```diff
@@ -160,27 +160,20 @@ def predict(self, docs: Iterable[Doc]) -> List[Ints2d]:
     def _scores2guesses(self, docs, scores):
         guesses = []
         for doc, doc_scores in zip(docs, scores):
-            if self.top_k == 1:
-                doc_guesses = doc_scores.argmax(axis=1).reshape(-1, 1)
-            else:
-                doc_guesses = np.argsort(doc_scores)[..., : -self.top_k - 1 : -1]
-
-            if not isinstance(doc_guesses, np.ndarray):
-                doc_guesses = doc_guesses.get()
-
+            if not isinstance(doc_scores, np.ndarray):
+                doc_scores = doc_scores.get()
```

Comment:
To be honest I can't say much about either question — why we are pinning to the CPU or why we are doing it this way. I just copied the existing code for `doc_guesses` (for which there might conceivably be different factors at play as it's a smaller array). @danieldk can you shed any light on these questions?

Reply:
What I also still have to investigate is whether or not performing the `argmax` in a loop causes a measurable speed penalty when using a GPU. If so, I'll change the method back to use separate strategies for `argmax = 1` and `argmax > 1`.

Reply:
We are iterating over the scores here. Doing that on the GPU kills performance.

Reply:
> Though, it does lead to the degenerate O(n^2) case when someone says 'lets try all edit trees'. 

An important point is that even if you say "let's try all edit trees" it only iterates through them until it comes across a tree that can be applied to the raw token text. And in any normal model a good proportion of trees are e.g. 'add a `u` to the end' which can be applied to any text. This will be shown in the results I'm collecting: increasing the value of `top_k` beyond about 5 has no significant effect, and increasing it beyond about 10 has no effect, because it's so unlikely that there is no applicable tree in the first handful tried.

Reply:
Having collected the first results on GPU (I'd been focussed on CPU until now) I can confirm that:

- the speed problem that this PR fixes relates exclusively to CPU and is not relevant to GPU;
- the performance on GPU with this PR is either equivalent or slightly worse than the performance with the pre-existing code. Depending on what the numbers I'm collecting show, it might make sense to retain the pre-existing code for GPU and use the new code on CPU;
- What is definitely the case: if you don't pin `doc_scores` to the CPU when working on the GPU, inference is thirteen times slower than if you work on the GPU and do copy it across and three times slower than if you work exclusively on the CPU. So pinning is definitely necessary if we end up using the new code for GPU.

Reply:
> What is definitely the case: if you don't pin doc_scores to the CPU when working on the GPU, inference is thirteen times slower than if you work on the GPU and do copy it across and three times slower than if you work exclusively on the CPU. So pinning is definitely necessary if we end up using the new code for GPU.

Yeah, my first week of GPU performance improvements in spaCy/Thinc was finding loops that iterate over GPU arrays by element. These 4-byte transfers are insanely expensive.